website/content/docs/concepts/request-limiter.mdx

```diff
@@ -0,0 +1,71 @@
+---
+layout: docs
+page_title: 'Request Limiter'
+description: >-
+  Vault provides an adaptive concurrency limiter to protect the Vault server
+  from overload.
+---
+
+# Request Limiter
+
+@include 'alerts/enterprise-only.mdx'
+
+@include 'alerts/beta.mdx'
+
+This document contains conceptual information about the **Request Limiter** and
+its user-facing effects.
+
+## Preventing overload
+
+The Request Limiter aims to prevent overload by proactively detecting latency
+and adapting the number of allowed in-flight requests.
+
+This is done in two phases at the beginning of an HTTP request:
+
+1. Consult the current number of allowed in-flight requests. If the new request
+would exceed this limit, immediately reject it, indicating that the client
+should retry later.
+
+2. If the request is allowed, begin a measurement of its latency, allowing the
+Request Limiter to calculate a new limit.
+
+## Resource constraints
+
+The Request Limiter intentionally focuses on preventing overload derived from
+resource-constrained operations on the Vault server. There are two known
+resource-constrained request paths in Vault:
+
+1. Back pressure from the storage backend, resulting in bufferbloat in the WAL
+system (enterprise) for `Write`-based HTTP methods.
+
+2. Back pressure from CPU over-utilization via PKI issue requests (specifically
+for RSA keys), resulting in failed heartbeats.
+
+Storage constraints can be accounted for by limiting logical requests according
+to their `http.Method`. We only limit requests with write-based methods, since
+these will result in storage Puts and exhibit the aforementioned bufferbloat.
+
+CPU constraints are accounted for using the same underlying library and
+technique; however, they require special treatment. The maximum number of
+concurrent pki/issue requests found in testing (again, specifically for RSA
+keys) is far lower than the minimum tolerable write request rate.
+
+In both cases, utilization will be effectively throttled before Vault reaches
+any degraded state. The resulting `503 - Service Unavailable` is a retryable
+HTTP response code, which can be handled to gracefully retry and eventually
+succeed.  Clients should handle this by retrying with jitter and exponential
+backoff. This is done within Vault's API, using the go-retryablehttp library.
+
+## Read requests
+
+HTTP methods, such as `GET` and `LIST` are currently not subject to request
+limiting. This allows operators to continue querying server state without
+needing to retry. This decision was based on the observation that read-based
+requests are not as likely to cause untenable load than their
+resource-constrained counterparts.
```

Comment:
```suggestion
needing to retry.
```

I'm not sure this statement is accurate - we probably will want to limit read requests in the future too since every request type will hit resource constraints and cause instability eventually. The key is that it's a _different resource constaint_ such as memory usage for open connections, or CPU usage with too many goroutines.

We already said above that the write limiter doesn't bother with read requests because those don't hit the same bottleneck and have different performance characteristics so it seems fine to leave this out here and in future we can replace this with additional limiter types as we choose to tackle different sources of resource saturation.

Reply:
Removed in https://github.com/hashicorp/vault/pull/25557/commits/7a38c1ff7ee0448f14b115478f967947a06e704e